src/plt-elf.c

```diff
@@ -224,54 +223,62 @@ static uintptr_t get_offset(struct rel_info *info, ElfW(Sym) *symtab,
     return 0;
 }
 
-static plt_fn **plt_get_offset(plt_ctx ctx, plt_lib lib, const char *name)
+static size_t get_offsets(plt_ctx ctx, plt_lib lib, const char *name, plt_offset ** offsets)
 {
-    ElfW(Sym) *symtab   = (ElfW(Sym)*)  lib_dt_lookup(lib, DT_SYMTAB);
-    const char *strtab  = (const char*) lib_dt_lookup(lib, DT_STRTAB);
-    ElfSWord type = (ElfSWord) lib_dt_lookup_val(lib, DT_PLTREL);
-
-    ElfW(Rel)   *jmprel = lib_dt_lookup(lib, DT_JMPREL);
-    ElfW(Rel)   *rel    = lib_dt_lookup(lib, type);
-    ElfWord rel_sz = lib_dt_lookup_val(lib, DT_PLTRELSZ);
-    ElfWord relent_sz = lib_dt_lookup_val(lib, type + 2);
+  ElfW(Sym) *symtab   = (ElfW(Sym)*)  lib_dt_lookup(lib, DT_SYMTAB);
+  const char *strtab  = (const char*) lib_dt_lookup(lib, DT_STRTAB);
 
-    if (!symtab || !strtab || !type || !(rel || jmprel) || !rel_sz || !relent_sz)
-        return NULL;
+  ElfW(Rel)   *jmprel = lib_dt_lookup(lib, DT_JMPREL);
+  ElfW(Rel)   *rel    = lib_dt_lookup(lib, DT_RELA);
+  ElfWord rel_sz = lib_dt_lookup_val(lib, DT_RELASZ);
+  ElfWord jmprel_sz = lib_dt_lookup_val(lib, DT_PLTRELSZ);
+  ElfWord relent_sz = lib_dt_lookup_val(lib, DT_RELAENT);
 
-    struct rel_info info = {
-        .tab = rel,
-        .size = rel_sz,
-        .entry_sz = relent_sz,
-    };
-
-    uintptr_t off = get_offset(&info, symtab, strtab, name);
-    if (!off) {
-        /* relocation might be in JMPREL (e.g. .rela.plt) instead */
-        info.tab = jmprel;
-        off = get_offset(&info, symtab, strtab, name);
-    }
+  if (!symtab || !strtab || !(rel || jmprel) || !(rel_sz || jmprel_sz) || !relent_sz)
+    return 0;
 
-    ElfW(Addr) base = (ElfW(Addr)) lib->l_addr;
+  ElfW(Addr) base = (ElfW(Addr)) lib->l_addr;
 #ifdef __FreeBSD__
-    if (lib == ctx->r_map)
-        base = 0;
+  if (lib == ctx->r_map)
+    base = 0;
 #endif
 
-    if (off)
-        return (plt_fn **) (off + base);
-    return NULL;
+  size_t n = 0;
+  if (offsets) {
+    *offsets = NULL;
+  }
+
+  struct rel_info info[] = {
+  {
+    .tab = rel,
+    .size = rel_sz,
+    .entry_sz = relent_sz,
+  },
+  {
+    .tab = jmprel,
+    .size = jmprel_sz,
+    .entry_sz = relent_sz,
+  }};
+
+  for (size_t i = 0; i < sizeof(info) / sizeof(struct rel_info); ++i) {
+    uintptr_t off = get_offset(&info[i], symtab, strtab, name);
+    if (off) {
+      if (offsets) {
+        *offsets = mmk_realloc(*offsets, (n + 1) * sizeof(plt_offset));
+        (*offsets)[n] = (plt_offset) { .offset = (plt_fn **)(base + off) };
+      }
+      ++n;
+    }
+  }
+
+  return n;
 }
 
 plt_offset *plt_get_offsets(plt_ctx ctx, plt_lib lib, const char *name, size_t *n)
 {
-    plt_fn **off = plt_get_offset(ctx, lib, name);
-    if (off) {
-        plt_offset *ot = mmk_malloc(sizeof (plt_offset));
-        *n = 1;
-        *ot = (plt_offset) { .offset = off };
-        return ot;
-    }
-    return NULL;
+    plt_offset *ot = NULL;
+    *n = get_offsets(ctx, lib, name, &ot);
+    return ot;
```

Comment:
Since `plt_get_offsets` pretty much just becomes useless, I would just rename `get_offsets` to `plt_get_offsets` and have it return the allocated offset array rather than retrieving it by pointer.

Reply:
I could, but `plt_get_lib` still needs a way check for symbol existence [here](https://github.com/Snaipe/Mimick/blob/7681d1ee8e0bffff806751092947648a6bebff24/src/plt-elf.c#L198) and I'd rather not incur in a memory allocation if there's no use for it. Thus this two functions. WDYT?

Reply:
Ah, I see. That makes sense, this is fine as-is.